ci/src/mainnet_revisions/mainnet_revisions.py

```diff
@@ -0,0 +1,189 @@
+
+
+def get_saved_nns_subnet_version(repo_root: pathlib.Path):
+    """Get the last known version running on the NNS subnet."""
+    saved_versions = get_saved_versions(repo_root=repo_root)
+    return saved_versions.get("subnets", {}).get(nns_subnet_id, "")
+
+
+def get_saved_app_subnet_version(repo_root: pathlib.Path):
+    """Get the last known version running on an App subnet."""
+    saved_versions = get_saved_versions(repo_root=repo_root)
+    return saved_versions.get("subnets", {}).get(app_subnet_id, "")
+
+def get_subnet_replica_version(subnet_id: str) -> str:
+    ic_admin_path = os.getenv("IC_ADMIN_PATH", "ic-admin")
+    output = subprocess.run(
+        [ic_admin_path, "--nns-urls", "https://ic0.app" ,"get-subnet", subnet_id],
+        text=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE
+    )
+
+    if output.returncode != 0:
+        print("Failed to get subnet", subnet_id, "Error:", output.stderr)
+        exit(1)
+
+    response = json.loads(output.stdout)
+    return response["records"][0]["value"]["replica_version_id"]
+
+def get_repo_root() -> str:
+    return subprocess.run(
+        ["git", "rev-parse", "--show-toplevel"],
+        text=True,
+        stdout=subprocess.PIPE
+    ).stdout.strip()
+
+def main():
+    """Do the main work."""
+
+    class HelpfulParser(argparse.ArgumentParser):
+        """An argparse parser that prints usage on any error."""
+
+        def error(self, message):
+            sys.stderr.write("error: %s\n" % message)
+            self.print_help()
+            sys.exit(2)
+
+    parser = HelpfulParser()
+
+    parser.add_argument("--verbose", "-v", action="store_true", help="Verbose mode")
+
+    args = parser.parse_args()
+
+    if args.verbose:
+        logging.basicConfig(level=logging.DEBUG)
+    else:
+        logging.basicConfig(level=logging.INFO)
+
+    current_nns_version = get_subnet_replica_version(nns_subnet_id)
+    logging.info("Current NNS subnet (%s) revision: %s", nns_subnet_id, current_nns_version)
+    current_app_subnet_version = get_subnet_replica_version(app_subnet_id)
+    logging.info("Current App subnet (%s) revision: %s", app_subnet_id, current_app_subnet_version)
+
+    ic_repo_push_token = os.environ.get("PUSH_TO_IC", "ENV_VAR_NOT_SET")
+    repo = "dfinity/ic"
+    remote_url = f"https://oauth2:{ic_repo_push_token}@github.com/{repo}.git"
+
+    repo_root = pathlib.Path(get_repo_root())
+
+    running_on_ci = os.environ.get("CI") or os.environ.get("GITHUB_ACTION")
+    if running_on_ci:
+        logging.info("Setting up git credentials for CI")
+        subprocess.check_call(["git", "config", "--global", "credential.helper", "store --file ~/.git-credentials"])
+        with open(os.path.expanduser("~/.git-credentials"), "w") as f:
+            f.write(f"{remote_url}\nusername=oauth2\npassword={ic_repo_push_token}\n")
+    else:
+        logging.info("Running locally, using default git credentials of the current user")
```

Comment:
See my comment about persisting the token in the git config using actions/checkout above. With that I don't think you need this:
```suggestion
```